src/Effect/Ref.purs

```diff
@@ -2,6 +2,38 @@
 -- |
 -- | _Note_: `Control.Monad.ST` provides a _safe_ alternative to `Ref` when
 -- | mutation is restricted to a local scope.
+-- |
+-- | You'll notice that all of the functions that operate on a `Ref`
+-- | (e.g. `new`, `modify`, `modify_`, `read`) return an `Effect (Ref s)`
+-- | and not a `Ref s`. Here's what would happe if the `Ref s` was not
+-- | wrapped in an `Effect`:
+-- | 1. In some situations, one could violate referential transparency.
```

Comment:
I think putting it this way downplays the importance of Effect to retain referential transparency here; I'd argue that a Ref which could be read or written outside of Effect would end up violating referential transparency in almost all situations, since every time you write a new value to a ref, you change the result of reading from it. I also think we could do more to explain what referential transparency is, as if the reader isn't already familiar with the term, these docs probably won't make a lot of sense (and if they are familiar with it, they likely understand why reading or writing a Ref is effecful already anyway).

Reply:
> I think putting it this way downplays the importance of Effect to retain referential transparency here

So... change the wording to "In all situations, one will violate referential transparency" ?

> I also think we could do more to explain what referential transparency is

Odd. Usually you want docs to explain less and have readers look up terms elsewhere. Why the change here?

Reply:
> change the wording to “in all situations...”?

Well, that’s not quite accurate either - it’s definitely possible that a Ref is initialised and then never written to, in which case all of the reads would return the same value and you wouldn’t observe a violation.

> Why the change here?

 I want docs to only address things that are directly relevant to the topic at hand; in this case, referential transparency is directly relevant to why refs are effectful. I also expect that it only needs a sentence or two; if it needed more than that, I’d agree that it should go elsewhere. If you can point to an example where you think I’m being inconsistent I’m happy to reconsider.